test/integration/job/job_test.go

```diff
@@ -1026,6 +1029,202 @@ func TestIndexedJob(t *testing.T) {
 	validateTerminatedPodsTrackingFinalizerMetric(t, 5)
 }
 
+func TestElasticIndexedJob(t *testing.T) {
+	type jobUpdate struct {
+		completions          int
+		succeededIndexes     []int
+		failedIndexes        []int
+		wantSucceeded        int
+		wantSucceededIndexes string
+		wantFailed           int
+	}
+	var initialCompletions = 3
+
+	cases := map[string]struct {
+		featureGate bool
+		jobUpdates  []jobUpdate
+	}{
+		"feature flag off, mutation not allowed": {
+			jobUpdates: []jobUpdate{
+				{
+					completions: 4,
+				},
+			},
+		},
+		"scale up, verify that the range expands, and the job finishes successfully when indexes including the ones in the new range exit successfully": {
+			featureGate: true,
+			jobUpdates: []jobUpdate{
+				{
+					// Scale up completions 3->4 then succeed indexes 0-3
+					completions:          4,
+					succeededIndexes:     []int{0, 1, 2, 3},
+					wantSucceeded:        4,
+					wantSucceededIndexes: "0-3",
+					wantFailed:           0,
+				},
+			},
+		},
+		"scale down, verify that the range shrinks, and the job finishes successfully when indexes only in the smaller range finishes successfully, and verify that failures that happened for indexes that are now outside the range still count": {
+			jobUpdates: []jobUpdate{
+				// First succeed index 1 and fail index 2 while completions is still original value (3).
```

Comment:
how are those tests succeeding without setting the feature gate to true?

Reply:
They were succeeding due to a bug, basically if the `updateJob` call returned an error, the test would check if `tc.featureGate` is false, and if so continue without error (basically treating `tc.featureGate=false` as expecting an error i.e. test passing. So the validation was working correctly and returning an error but the error was incorrectly being treated as expected so the tests passed.

Now that I have updated the test case struct to have a `wantErr` with the specific errors expected instead, and use that to determine if an error is expected or not, this issue is now fixed.